Add a flag to disable force detach behavior in kube-controller-manager

rohitssingh opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR adds a flag to allow the 6-minute force detach window to be disabled.

With this flag enabled, `ControllerUnpublishVolume` RPCs will not be dispatched if a given pod is stuck unmounting for 6-minutes and the node is deemed to be Unhealthy in that window.

#### Which issue(s) this PR fixes:

Fixes #120328

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
This change adds the following CLI option for `kube-controller-manager`:
* `disable-force-detach` (defaults to `false`): Prevent force detaching volumes based on maximum unmount time and node status. If enabled, the non-graceful node shutdown feature must be used to recover from node failure (see https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/). If enabled and a pod must be forcibly terminated at the risk of corruption, then the appropriate VolumeAttachment object (see here: https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/) must be deleted.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

rohitssingh committed:
Add "disable-force-detach-on-timeout" flag to kube-controller-manager

rohitssingh commented:
@bswartz @jingxu97 @msau42 @saad-ali

jingxu97 commented on pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:
```diff
@@ -75,7 +75,7 @@ func Test_Run_Positive_DoNothing(t *testing.T) {
 	nodeLister := informerFactory.Core().V1().Nodes().Lister()
 	reconciler := NewReconciler(
-		reconcilerLoopPeriod, maxWaitForUnmountDuration, syncLoopPeriod, false, dsw, asw, ad, nsu, nodeLister, fakeRecorder)
+		reconcilerLoopPeriod, maxWaitForUnmountDuration, syncLoopPeriod, false, false, dsw, asw, ad, nsu, nodeLister, fakeRecorder)
```
wondering whether you could set a global var instead of adding false in all places. In the future if this is changed, you only need to change one place instead?

msau42 replied:
global variable is not good for unit tests that can run in parallel. Another option is to refactor the `NewReconciler` function to take an options struct with default values. But I think that is out of scope for this change.

rohitssingh replied:
Thanks; the I'm marking this as resolved for now.

alexzielenski commented:
/remove-sig api-machinery
/sig apps
/sig storage

carlory commented:
/kind feature
/ok-to-test

carlory commented on cmd/kube-controller-manager/app/options/attachdetachcontroller.go:
```diff
@@ -37,6 +37,7 @@ func (o *AttachDetachControllerOptions) AddFlags(fs *pflag.FlagSet) {
 
 	fs.BoolVar(&o.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile-sync", false, "Disable volume attach detach reconciler sync. Disabling this may cause volumes to be mismatched with pods. Use wisely.")
 	fs.DurationVar(&o.ReconcilerSyncLoopPeriod.Duration, "attach-detach-reconcile-sync-period", o.ReconcilerSyncLoopPeriod.Duration, "The reconciler sync wait time between volume attach detach. This duration must be larger than one second, and increasing this value from the default may allow for volumes to be mismatched with pods.")
+	fs.BoolVar(&o.DisableForceDetachOnUnmount, "disable-force-detach", false, "Prevent force detaching volumes based on maximum unmount time and node status. If enabled, the non-graceful node shutdown feature must be used to recover from node failure (see https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/). If enabled and a pod must be forcibly terminated at the ris ...
```
The field name isn't consistent with the flag name.

rohitssingh replied:
Thanks; resolved.

rohitssingh commented:
/retest

alexzielenski commented:
/remove-sig api-machinery

msau42 commented:
/label api-review

liggitt commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -215,8 +218,8 @@ func (rc *reconciler) reconcile(ctx context.Context) {
-			// Force detach volumes from unhealthy nodes after maxWaitForUnmountDuration.
-			forceDetach := !isHealthy && timeout
+			// Force detach volumes from unhealthy nodes after maxWaitForUnmountDuration if force detach is enabled
+			forceDetach := !isHealthy && timeout && !rc.disableForceDetachOnUnmount
```
does the disableForceDetach bit need to inform `verifySafeToDetach := !(timeout || hasOutOfServiceTaint)` below as well?

msau42 replied:
It wouldn't make it past L231 but it won't hurt to add in case this function ever gets refactored in the future

jingxu97 replied:
Whether to trigger detach and whether to verifySafeToDetach can be affected by 5 variables:
isHealthy | timeout | disableForceDetachOnUnmount | MountedByNode | hasOutOfServiceTaint

I feel it brings some trouble when reasoning about all the conditions. Trying to think of a way to better express this.

```
verifySafeToDetach = true
if hasOutOfServiceTaint {
       // this is a sufficient condition to continue with detaching without verification no matter any other conditions
       verifySafeToDetach = false
} else if  timeout & !isHealthy & !disableForceDetachOnTimeOut {
       // it is already timeout and node is NOT healthy, and forceDetach is NOT disabled, force detach can happen
      verifySafeToDetach = false
} else if !MountedByNode {
       // it is no longer mounted, detach can be triggered.
} else {
     // all other condition, we should skip detach
     continue
}

... detach operation begins ...

```

rohitssingh replied:
I've made the following changes based on the feedback here:
* Added a logging message for when the timeout expires but is suppressed by the flag.
* Updated the semantics of how the `timeout` variable is calculated; it now includes `rc. disableForceDetachOnTimeOut`. This is because when I added the unit test to bypass this flag with the node taint, we were updating the wrong counter in the code below (the detach was counting towards `metrics.RecordForcedDetachMetric(metrics.ForceDetachReasonTimeout)` and not `metrics.RecordForcedDetachMetric(metrics.ForceDetachReasonOutOfService)`). This means that `rc. disableForceDetachOnTimeOut` is also included in `verifySafeToDetach`.
* @jingxu97: The more that I stare at this code, the more I think your refactor makes sense. Is this something I can pursue as a follow-up change?

jingxu97 replied:
lgtm, thank you!

rohitssingh replied:
Marking this as resolved; will follow up later.

liggitt commented on cmd/kube-controller-manager/app/options/attachdetachcontroller.go:
```diff
@@ -37,6 +37,7 @@ func (o *AttachDetachControllerOptions) AddFlags(fs *pflag.FlagSet) {
 
 	fs.BoolVar(&o.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile-sync", false, "Disable volume attach detach reconciler sync. Disabling this may cause volumes to be mismatched with pods. Use wisely.")
 	fs.DurationVar(&o.ReconcilerSyncLoopPeriod.Duration, "attach-detach-reconcile-sync-period", o.ReconcilerSyncLoopPeriod.Duration, "The reconciler sync wait time between volume attach detach. This duration must be larger than one second, and increasing this value from the default may allow for volumes to be mismatched with pods.")
+	fs.BoolVar(&o.DisableForceDetachOnUnmount, "disable-force-detach-on-unmount", false, "Prevent force detaching volumes based on maximum unmount time and node status. If enabled, the non-graceful node shutdown feature must be used to recover from node failure (see https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/). If enabled and a pod must be forcibly terminated ...
```
nit: two urls in the flag help text is probably too long... is there a single short URL we can point to here?

msau42 commented on cmd/kube-controller-manager/app/options/attachdetachcontroller.go:
```diff
@@ -37,6 +37,7 @@ func (o *AttachDetachControllerOptions) AddFlags(fs *pflag.FlagSet) {
 
 	fs.BoolVar(&o.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile-sync", false, "Disable volume attach detach reconciler sync. Disabling this may cause volumes to be mismatched with pods. Use wisely.")
 	fs.DurationVar(&o.ReconcilerSyncLoopPeriod.Duration, "attach-detach-reconcile-sync-period", o.ReconcilerSyncLoopPeriod.Duration, "The reconciler sync wait time between volume attach detach. This duration must be larger than one second, and increasing this value from the default may allow for volumes to be mismatched with pods.")
+	fs.BoolVar(&o.DisableForceDetachOnUnmount, "disable-force-detach-on-unmount", false, "Prevent force detaching volumes based on maximum unmount time and node status. If enabled, the non-graceful node shutdown feature must be used to recover from node failure (see https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/). If enabled and a pod must be forcibly terminated ...
```
Maybe I would call the flag "DisableForceDetachOnTimeout" to better reflect the conditions

rohitssingh replied:
Thanks; I've refactored the code accordingly.

msau42 commented on pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:
```diff
@@ -1111,6 +1111,110 @@ func Test_Run_OneVolumeDetachOnUnhealthyNode(t *testing.T) {
+	waitForDetachCallCount(t, 0 /* expectedDetachCallCount */, fakePlugin)
+
+	// Act
+	// Mark the node unhealthy
```
It would be good to also add the taint and validate that detach does go through.

rohitssingh replied:
Done.

That was a very worthwhile test to add; it helped me realize that the wrong counter was being updated on detach as I mentioned above.

Checking the metrics (which are stashed in a global variable) was not great though.

Marking this as resolved.

rohitssingh commented:
/retest

jiahuif commented:
/remove-sig api-machinery

carlory commented on cmd/kube-controller-manager/app/options/attachdetachcontroller.go:
```diff
@@ -37,6 +37,7 @@ func (o *AttachDetachControllerOptions) AddFlags(fs *pflag.FlagSet) {
 
 	fs.BoolVar(&o.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile-sync", false, "Disable volume attach detach reconciler sync. Disabling this may cause volumes to be mismatched with pods. Use wisely.")
 	fs.DurationVar(&o.ReconcilerSyncLoopPeriod.Duration, "attach-detach-reconcile-sync-period", o.ReconcilerSyncLoopPeriod.Duration, "The reconciler sync wait time between volume attach detach. This duration must be larger than one second, and increasing this value from the default may allow for volumes to be mismatched with pods.")
+	fs.BoolVar(&o.DisableForceDetachOnTimeout, "disable-force-detach-on-timeout", false, "Prevent force detaching volumes based on maximum unmount time and node status. If this flag is set to true, the non-graceful node shutdown feature must be used to recover from node failure (see https://t.ly/Gl1qo). If this flag is set to true and a pod must be forcibly terminated at the risk of corruption, then ...
```
What's the `https://t.ly` ? It's not a kubernetes official link. And I don't think it's necessary to add this link in the description because the non-graceful node shutdown feature is enough to users. 


> If this flag is set to true and a pod must be forcibly terminated at the risk of corruption, then the appropriate VolumeAttachment object (see here: https://t.ly/-atgq) must be deleted."

It seems to explain the details of the non-graceful node shutdown feature. I think it's not necessary to add because it doesn't relate to the flag itself. If users wants to know the details, they can search it in the kubernetes official document.

How about it?

```
Prevent force detaching volumes when maximum unmount timeout is reached. If true, the non-graceful node shutdown feature should be used to recover from node failure.
```

rohitssingh replied:
Thanks for the feedback.

To provide some background: I got here because while I was testing this change in our internal repo, I got a review comment from @jingxu97 to include a link to the non-graceful shutdown node feature to help users understand the proper workflow to address hung unmounts if this feature is enabled:

```
If enabled, the non-graceful node shutdown feature could be used to recover from node failure? And add the feature blog link?
Besides non-graceful node shutdown, there is other manual way to just detach volume in case non-graceful node shutdown does not work in certain scenarios.
```

So I added in the two k8s links to resolve those comments:
* Information on non-graceful node shutdown: https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/
* Information on the CR to remove for manual remediation: https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/

I then received feedback from @liggitt indicating that the URLs were making things too long:

```
nit: two urls in the flag help text is probably too long... is there a single short URL we can point to here?
```

I was hoping that if I used a URL shortener like http://t.ly, I'd still be able to keep the two links that I had been earlier asked to include.

And so I replaced the two links with shorter versions:
* https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/ -> https://t.ly/Gl1qo
* https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/ -> https://t.ly/-atgq

@carlory, @jingxu97, @liggitt: Do you folks have any suggestions on how to best reconcile this feedback?

I'm relatively amenable to whatever we decide, but I would like to ensure everyone's concerns are addressed.

rohitssingh replied:
One question that comes to mind: Is there a preferred URL shortener for the kubernetes project that I should be using in lieu of https://t.ly?

liggitt replied:
https://github.com/kubernetes/website/blob/main/static/_redirects allows creating redirects for the k8s.io domain, so `http://k8s.io/nodes#non-graceful-node-shutdown` could redirect to https://k8s.io/nodes/#non-graceful-node-shutdown

I'd suggest having a single URL in the flag help and making sure the content of that page linked out to relevant APIs or pages.

rohitssingh replied:
Thank for the clarification @liggitt.

I spent some time poking around the website code to find where to put this.

I think it makes sense to create a small entry under https://github.com/kubernetes/website/tree/main/content/en/docs/tasks, maybe underneath a new `storage` subfolder, called `disable-force-detach-on-timeout` where we point to https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/ (the preferred remediation if unmount is stuck due to this flag) and https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/ (the CR that has to be deleted if the preferred remediation doesn't work or isn't wiling to taint the entire node).

Does this sound like the right way to proceed? If so, I'm hoping I can create a new commit for https://github.com/kubernetes/website with:
* A new page with a very short description of what this flag is and links to the other pages
* A shortlink to the new page to include into the description of this commit

I'll then attach the new commit to this PR assuming I can have commits from different repositories attached to the same PR. If that doesn't work, I'll try creating a new PR.

Please feel free to let me know if I have misunderstood something or if you disagree with this proposed course of action.

Thanks!

rohitssingh replied:
Apologies for the late response; I was pulled into another issue.

Thanks to some great suggestions from @bswartz, I was able to put together a documentation change here: https://github.com/kubernetes/website/pull/44931

The short redirect link I've picked is `/docs/node-disable-force-detach-on-timeout/` which maps to a short section I've added to `/docs/concepts/architecture/nodes/#disable-force-detach-on-timeout`.

The proposed changes and redirect can currently be tested here:
https://deploy-preview-44931--kubernetes-io-main-staging.netlify.app/docs/node-disable-force-detach-on-timeout/; this currently redirects to https://deploy-preview-44931--kubernetes-io-main-staging.netlify.app/docs/concepts/architecture/nodes/#disable-force-detach-on-timeout.

rohitssingh replied:
Oh, and the documentation PR can be found here: https://github.com/kubernetes/website/pull/44931

rohitssingh replied:
I spoke to @bswartz; I think we should be good to go with this change?

@carlory, @liggitt, @msau42: Do the proposed URLs make sense?

In this case, it would be: https://k8s.io/docs/node-disable-force-detach-on-timeout/

If so, does it make sense to commit this before the other documentation change goes in (since I cannot commit both together)? How should we stage this?

rohitssingh replied:
Note: Based on review feedback, the new link will be https://k8s.io/docs/storage-disable-force-detach-on-timeout/

@carlory, @liggitt, @msau42: Please feel free to let me know if this looks okay; you can see a preview of the page here: https://deploy-preview-45119--kubernetes-io-vnext-staging.netlify.app/docs/storage-force-detach-on-timeout

sftim replied:
Feedback from SIG Docs: can we use a different short URL?

(I / we didn't know folks were planning to mint one)

See https://github.com/kubernetes/website/pull/45119/files#r1539091498

rohitssingh commented:
/remove-lifecycle stale

I should be pushing out a change shortly...

seans3 commented:
/remove-sig api-machinery

seans3 commented:
/triage accepted

YuikoTakada commented:
Hi, thank you for your PR. I understand your `why we need it` and this issue should be fixed. 
I don't think this issue must be fixed only by `flag` . I think you issue can be fixed by allowing to configure `maxWaitForUnmountDurationwhich` reported by https://github.com/kubernetes/kubernetes/issues/68086 .
How about making `maxWaitForUnmountDurationwhich` configurable and if minus value is passed to `maxWaitForUnmountDurationwhich` , make `maxWaitForUnmountDurationwhich` infinity (which means disable force detach behavior) ?

rohitssingh commented:
> Hi, thank you for your PR. I understand your `why we need it` and this issue should be fixed. I don't think this issue must be fixed only by `flag` . I think you issue can be fixed by allowing to configure `maxWaitForUnmountDurationwhich` reported by #68086 . How about making `maxWaitForUnmountDurationwhich` configurable and if minus value is passed to `maxWaitForUnmountDurationwhich` , make `maxWaitForUnmountDurationwhich` infinity (which means disable force detach behavior) ?

@YuikoTakada: Thanks for the feedback.

I think there is merit to your idea. However, I think we would like to proceed with this PR as-is for the following reasons:
* As I have mentioned, this bug has caused a number of data corruption issues in our storage systems, and it has taken a significant effort to get this PR to this point. In the past two weeks, we have encountered yet another spate of corruption issues, which has increased the urgency of getting this change in soon. I am worried that if we change the nature of this setting yet again, we likely will have to go through review again and make more changes to the associated k8s website review. I feel like we are close to establishing some consensus on this front on the form that this change should take, and I think we should proceed with what we have here.
* Strictly from my team's perspective, there really is no value in having a higher timeout. Corruption is corruption, and from our tests and our usecases the presence of this behavior is totally intolerable. In other words, there is no world where a 6-minute timeout is undesirable, but a 60-minute timeout is okay or a 120-minute timeout is okay. We just really need a way to make this behavior to stop.

YuikoTakada commented:
Thank you for your kindly detailed explanation. I agree with your chage.
It's just out of interest, in your team, the unmount of a pod takes too long time in spite of node broken? Approximately how many minutes does it take to unmount?

msau42 commented:
/lgtm
/approve

msau42 commented:
/assign @jpbetz

bswartz commented:
/lgtm

jpbetz commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -208,14 +211,19 @@ func (rc *reconciler) reconcile(ctx context.Context) {
-			// Force detach volumes from unhealthy nodes after maxWaitForUnmountDuration.
+			// Force detach volumes from unhealthy nodes after maxWaitForUnmountDuration if force detach is enabled
+			// Ensure that the timeout condition checks this correctly so that the correct metric is updated below
+			timeout := drainTimeoutExpired && !rc.disableForceDetachOnTimeout
```
```suggestion
			forceDetatchTimeoutExpired := drainTimeoutExpired && !rc.disableForceDetachOnTimeout
```

To clarify what this timeout is throughout this function (it's rather long)?

jpbetz commented on staging/src/k8s.io/kube-controller-manager/config/v1alpha1/types.go:
```diff
@@ -179,6 +179,9 @@ type AttachDetachControllerConfiguration struct {
 	// ReconcilerSyncLoopPeriod is the amount of time the reconciler sync states loop
 	// wait between successive executions. Is set to 60 sec by default.
 	ReconcilerSyncLoopPeriod metav1.Duration
+	// Flag to disable force detach if the maximum unmount time is exceeded.
+	// Is false by default, and thus force detach on unmount is enabled.
```
```suggestion
	// disableForceDetachOnTimeout disables force detach when the maximum unmount time is exceeded.
	// Is false by default, and thus force detach on unmount is enabled.
```

Since, here, this is a component config field, not a flag.

@msau42 Does this follow API review guidelines?  I can't remember if want "disable" fields or if we prefer "enable" fields that's can be used to disable functionality.  Note that we already have `DisableAttachDetachReconcilerSync` as a sibling field.

rohitssingh commented:
@jpbetz: Thanks for the feedback! I think I've addressed your comments. Please feel free to take another look.
@msau42 & @bswartz: Looks like the `lgtm` tag has been removed because I uploaded a new patch? Please feel free to take another look.

jpbetz commented:
/lgtm

jpbetz commented:
/approve

rohitssingh commented:
> Thank you for your kindly detailed explanation. I agree with your chage.
> It's just out of interest, in your team, the unmount of a pod takes too long time in spite of node broken? Approximately how many minutes does it take to unmount?

Hi @YuikoTakada,

I think we've encountered a number of different scenarios with force detach:
- In some cases, we have had misbehaving pods on otherwise healthy nodes that were being migrated and just ignored shutdown signals. These pods would never shutdown on their own, and often resulted in Zombie LUNs and data corruption.
- In other cases, we have had problems with the CSI driver not being able to process `NodeUnpublishVolume` or `NodeUnstageVolume` in a timely fashion. In cases where kubelet was restarted in the wrong window, or the load on a node was too high, the node would fall into an unhealthy state, leading to Zombie LUNs and corruption.

I don't think in any case we would have been able to unmount within any given time window. IMHO, We would very much prefer shutdown to hang indefinitely so that we can easily find & fix these problems (as opposed to discovering data corruption days or weeks later).

sftim commented:
BTW, we can use Markdown in the changelog, including actual hyperlinks.
